src/planner.cc

```diff
@@ -310,201 +409,146 @@ double Planner::Impl::euclidian_speed(const struct AxisTarget *t) {
+
+    if (cfg_->synchronous) motor_ops_->WaitQueueEmpty();
+
+    if (segment->planned.accel) ret = motor_ops_->Enqueue(accel_command);
+    if (has_move && ret) ret = motor_ops_->Enqueue(move_command);
+    if (segment->planned.decel && ret) ret = motor_ops_->Enqueue(decel_command);
+
+    // We always keep one segment to keep track of the last
+    // position and aux values.
+    if (planning_buffer_.size() > 1)
+      planning_buffer_.pop_front();
+    else {
+      // Let's create a zero-steps segment to hold last position.
+      // We have only one segment, last speed is always 0.
+      const StepsAxesRegister last_pos = segment->target.position_steps;
+      *segment = {};
+      segment->target.position_steps = last_pos;
+      path_halted_ = true;
+      ret = false;
+    }
   }
-  subtract_steps(&move_command, accel_command);
-  has_move = subtract_steps(&move_command, decel_command);
-
-  if (cfg_->synchronous) motor_ops_->WaitQueueEmpty();
-
-  // Make sure each segment gets added in case we get aborted
-  if (has_accel) ret = motor_ops_->Enqueue(accel_command);
-  if (ret && has_move) ret = motor_ops_->Enqueue(move_command);
-  if (ret && has_decel) ret = motor_ops_->Enqueue(decel_command);
-
-  last_aux_bits_ = target_pos->aux_bits;
 
   return ret;
 }
 
-// If we have enough data in the queue, issue motor move.
-bool Planner::Impl::issue_motor_move_if_possible() {
-  bool ret = true;
-  if (planning_buffer_.size() >= 3) {
-    ret =
-      move_machine_steps(planning_buffer_[0],   // Current established position.
-                         planning_buffer_[1],   // Position we want to move to.
-                         planning_buffer_[2]);  // Next upcoming.
-    planning_buffer_.pop_front();
-  }
-  return ret;
-}
-
+// At this stage of the planning we receive a new position we want to reach and
+// a target feedrate. We always try to reach the highest possible speed whenever
+// we move. For this reason, the feedrate is just interpreted as a speed limit.
+// This function performs three fundamental steps:
+// 1: Create a new empty planning segment in the planning buffer to be filled
+// with a new AxisTarget. 2: Determine the defining axis of the newly pushed
+// AxisTarget and its limits considering
+//    the angle w.r.t the previous segment and maximum accelerations and speeds
+//    of the other axes.
+// 3: Update the planning buffer (UpdateMotionProfile) to enforce constrained
+// motion
+//    that will always end up at 0 speed.
+// 4: Enqueue the front of the planning buffer to the motors if certain
+//    conditions occur.
 bool Planner::Impl::machine_move(const AxesRegister &axis, float feedrate) {
   assert(position_known_);  // call SetExternalPosition() after DirectDrive()
+
+  // We always assume there's enough space to store a new segment and that we
+  // have at least one segment.
+  assert((planning_buffer_.size() < PLANNING_BUFFER_CAPACITY - 1) &&
```

Comment:
With the version at head, there is now a `capacity()` call, so if you rebase, you can formulate that as
```
 assert(planning_buffer_.size() < planning_buffer_.capacity() && !planning_buffer_.empty());
```

Reply:
(solved)